src/openforms/forms/models/form_variable.py

```diff
@@ -37,66 +38,62 @@
 USER_DEFINED = Q(source=FormVariableSources.user_defined)
 
 
-class FormVariableManager(models.Manager):
+class FormVariableManager(models.Manager["FormVariable"]):
     use_in_migrations = True
 
     @transaction.atomic
     def create_for_form(self, form: "Form") -> None:
         form_steps = form.formstep_set.select_related("form_definition")
 
         for form_step in form_steps:
-            self.create_for_formstep(form_step)
-
-    def create_for_formstep(self, form_step: "FormStep") -> list["FormVariable"]:
-        form_definition_configuration = form_step.form_definition.configuration
-        component_keys = [
-            component["key"]
-            for component in iter_components(
-                configuration=form_definition_configuration, recursive=True
-            )
-        ]
-        existing_form_variables_keys = form_step.form.formvariable_set.filter(
-            key__in=component_keys,
-            form_definition=form_step.form_definition,
-        ).values_list("key", flat=True)
-
-        form_variables = []
-        for component in iter_components(
-            configuration=form_definition_configuration, recursive=True
-        ):
-            if (
-                is_layout_component(component)
-                or component["type"] in ("content", "softRequiredErrors")
-                or component["key"] in existing_form_variables_keys
-                or component_in_editgrid(form_definition_configuration, component)
-            ):
+            self.synchronize_for(form_step.form_definition)
+
+    @transaction.atomic
+    def synchronize_for(self, form_definition: FormDefinition):
+        """
+        Synchronize the form variables for a given form definition.
+
+        This creates, updates and/or removes form variables related to the provided
+        :class:`FormDefinition` instance. It needs to be called whenever the Formio
+        configuration of a form definition is changed so that our form variables in each
+        form making use of the form definition accurately reflect the configuration.
+
+        Note that we *don't* remove variables related to other form definitions, as
+        multiple isolated transactions for different form definitions can happen at the
+        same time.
```

Comment:
For my understanding: did the deadlock issue occur when two admins were modifying two different forms that shared a reusable form definition? Is this fixed now because it no longer does the recoupling via celery?

Reply:
I think the problem was that the task was scheduled to run 60s later and other write actions were happening and multiple transactions were trying to lock the same records in the table and competing for resources. Celery is definitely a big factor in this issue.

Laurens also saw issues when saving different forms in the admin in different tabs, so you didn't even need multiple different users doing stuff at the same time, but one customer was apparently building forms with 4 different people so that will have manifested much more obviously.